cmd/client/client.go

```diff
@@ -0,0 +1,216 @@
+// Copyright 2022 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"encoding/base64"
+	"errors"
+	"flag"
+	"fmt"
+	"net"
+	"net/url"
+	"os"
+	"os/signal"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/Jigsaw-Code/outline-ss-server/client"
+	onet "github.com/Jigsaw-Code/outline-ss-server/net"
+	"github.com/op/go-logging"
+	"github.com/shadowsocks/go-shadowsocks2/socks"
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+var logger *logging.Logger
+
+func init() {
+	var prefix = "%{level:.1s}%{time:2006-01-02T15:04:05.000Z07:00} %{pid} %{shortfile}]"
+	if terminal.IsTerminal(int(os.Stderr.Fd())) {
+		// Add color only if the output is the terminal
+		prefix = strings.Join([]string{"%{color}", prefix, "%{color:reset}"}, "")
+	}
+	logging.SetFormatter(logging.MustStringFormatter(strings.Join([]string{prefix, " %{message}"}, "")))
+	logging.SetBackend(logging.NewLogBackend(os.Stderr, "", 0))
+	logger = logging.MustGetLogger("")
+}
+
+type serverConfig struct {
+	host   string
+	port   int
+	cipher string
+	secret string
+}
+
+func parseKey(k string) (serverConfig, error) {
+	u, err := url.Parse(k)
+	if err != nil {
+		return serverConfig{}, err
+	}
+
+	port, err := strconv.Atoi(u.Port())
+	if err != nil {
+		return serverConfig{}, fmt.Errorf("invalid port: %v", err)
+	}
+
+	secret := u.User.String()
+	if !strings.Contains(secret, ":") {
+		b, err := base64.StdEncoding.DecodeString(secret)
+		if err != nil {
+			b, err = base64.RawStdEncoding.DecodeString(u.User.String())
+			if err != nil {
+				return serverConfig{}, fmt.Errorf("invalid password in key: %v", err)
+			}
+		}
+		secret = string(bytes.TrimSpace(b))
```

Comment:
Nah we don't need it. 
For a while I kept seeing a '\n' in the string after decoding, realized it because I was generating the test ss:// keys using `echo "..." | base64 -d` without putting a `-n` for the echo. I forgot to remove this after I figure it out. Thanks for catching